ui/core/components/suggest_gems_action.ts

```diff
@@ -0,0 +1,627 @@
+import { IndividualSimUI } from '../individual_sim_ui.js';
+import { Player } from '../player.js';
+import { Sim } from '../sim.js';
+import { Gear } from '../proto_utils/gear.js';
+import { EquippedItem } from '../proto_utils/equipped_item.js';
+import { TypedEvent } from '../typed_event.js';
+import { Stats } from '../proto_utils/stats.js';
+import { GemColor, Stat, Profession, ItemSlot } from '../proto/common.js';
+
+interface GemCapsData {
+	gemId: number
+	statCaps: Stats
+}
+
+interface SocketData {
+	itemSlot: ItemSlot
+	socketIdx: number
+}
+
+abstract class GemOptimizer {
+	protected readonly player: Player<any>;
+	protected readonly sim: Sim;
+	protected readonly gemPriorityByColor: Record<GemColor, Array<GemCapsData>>; 
+	abstract metaGemID: number;	
+	static allGemColors: Array<GemColor> = [GemColor.GemColorRed, GemColor.GemColorYellow, GemColor.GemColorBlue];
+	epWeights!: Stats;
+	useJcGems!: boolean;
+	isBlacksmithing!: boolean;
+	numSocketedJcGems!: number;
+	jcUpgradePriority: Array<GemCapsData>;
+
+	static jcUpgradesById: Record<number, number> = {
+		40118: 42154,
+		40125: 42156,
+		40112: 42143,
+		40111: 42142,
+		40119: 36767,
+	};
+
+	constructor(simUI: IndividualSimUI<any>) {
+		this.player = simUI.player;
+		this.sim = simUI.sim;
+
+		// Initialize empty arrays of gem priorities for each socket color
+		this.gemPriorityByColor = {} as Record<GemColor, Array<GemCapsData>>;
```

Comment:
Cant this be a empty object by default on line 23?
```
protected readonly gemPriorityByColor: Record<GemColor, Array<GemCapsData>> = {};
```

Reply:
I wanted to initialize all the keys explicitly with empty Arrays, since I query the length elsewhere, such as on line 82.